## Use the dataset “Titanic “ in Pandas DataFrame.

## Implement a Regression model with multiple variables: independent variables “Age” and “Fare”; and “Survived “ as dependent variable.

In [1]:
import pandas as pd
import statsmodels.api as sm
import matplotlib.pyplot as plt
%matplotlib inline

/Users/sergi/anaconda/lib/python3.6/site-packages/statsmodels/compat/pandas.py:56: FutureWarning: The pandas.core.datetools module is deprecated and will be removed in a future version. Please use the pandas.tseries module instead.
  from pandas.core import datetools


In [3]:
df = pd.read_csv('/Users/sergi/Desktop/MBDA/Master/Data Science Foundations/Data Science Foundations_Git/Session_2/2_titanic_dataset.csv')

In [4]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [5]:
df = df[['Survived','Age', 'Fare']]

In [6]:
df.head()

Survived   Age     Fare
0         0  22.0   7.2500
1         1  38.0  71.2833
2         1  26.0   7.9250
3         1  35.0  53.1000
4         0  35.0   8.0500

### With a Constant

In [9]:
df.isnull()

Survived    Age   Fare
0       False  False  False
1       False  False  False
2       False  False  False
3       False  False  False
4       False  False  False
5       False   True  False
6       False  False  False
7       False  False  False
8       False  False  False
9       False  False  False
10      False  False  False
11      False  False  False
12      False  False  False
13      False  False  False
14      False  False  False
15      False  False  False
16      False  False  False
17      False   True  False
18      False  False  False
19      False   True  False
20      False  False  False
21      False  False  False
22      False  False  False
23      False  False  False
24      False  False  False
25      False  False  False
26      False   True  False
27      False  False  False
28      False   True  False
29      False   True  False
..        ...    ...    ...
861     False  False  False
862     False  False  False
863     False   True  False
864     False  False  False
865     False  False  False
866     False  False  False
867     False  False  False
868     False   True  False
869     False  False  False
870     False  False  False
871     False  False  False
872     False  False  False
873     False  False  False
874     False  False  False
875     False  False  False
876     False  False  False
877     False  False  False
878     False   True  False
879     False  False  False
880     False  False  False
881     False  False  False
882     False  False  False
883     False  False  False
884     False  False  False
885     False  False  False
886     False  False  False
887     False  False  False
888     False   True  False
889     False  False  False
890     False  False  False

[891 rows x 3 columns]

In [12]:
df.isnull().sum()

Survived      0
Age         177
Fare          0
dtype: int64

In [13]:
df = df.dropna(axis=0)

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 714 entries, 0 to 890
Data columns (total 3 columns):
Survived    714 non-null int64
Age         714 non-null float64
Fare        714 non-null float64
dtypes: float64(2), int64(1)
memory usage: 22.3 KB


In [18]:
X = df[['Age', 'Fare']]
y = df['Survived']

In [20]:
X = sm.add_constant(X)
model = sm.OLS(y, X).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               Survived   R-squared:                       0.083
Model:                            OLS   Adj. R-squared:                  0.080
Method:                 Least Squares   F-statistic:                     32.02
Date:                Mon, 04 Dec 2017   Prob (F-statistic):           4.84e-14
Time:                        00:51:12   Log-Likelihood:                -474.62
No. Observations:                 714   AIC:                             955.2
Df Residuals:                     711   BIC:                             969.0
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.4210      0.041     10.270      0.000       0.340       0.501
Age           -0.0035      0.001     -2.880      0.004      -0.006      -0.001
Fare           0.0026      0.000      7.708      0.000       0.002       0.003
==============================================================================
Omnibus:                       17.741   Durbin-Watson:                   1.911
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               95.470
Skew:                           0.396   Prob(JB):                     1.86e-21
Kurtosis:                       1.393   Cond. No.                         154.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### Without a Constant

In [21]:
X = df[['Age', 'Fare']]
y = df['Survived']

In [22]:
X = sm.add_constant(X)
X["const"] = X["const"].apply(lambda x: x*0)
model = sm.OLS(y, X).fit()
model.summary()

/Users/sergi/anaconda/lib/python3.6/site-packages/statsmodels/regression/linear_model.py:1471: RuntimeWarning: divide by zero encountered in double_scalars
  return np.sqrt(eigvals[0]/eigvals[-1])
/Users/sergi/anaconda/lib/python3.6/site-packages/statsmodels/base/model.py:1036: RuntimeWarning: invalid value encountered in true_divide
  return self.params / self.bse
/Users/sergi/anaconda/lib/python3.6/site-packages/scipy/stats/_distn_infrastructure.py:879: RuntimeWarning: invalid value encountered in greater
  return (self.a < x) & (x < self.b)
/Users/sergi/anaconda/lib/python3.6/site-packages/scipy/stats/_distn_infrastructure.py:879: RuntimeWarning: invalid value encountered in less
  return (self.a < x) & (x < self.b)
/Users/sergi/anaconda/lib/python3.6/site-packages/scipy/stats/_distn_infrastructure.py:1818: RuntimeWarning: invalid value encountered in less_equal
  cond2 = cond0 & (x <= self.a)


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               Survived   R-squared:                       0.374
Model:                            OLS   Adj. R-squared:                  0.373
Method:                 Least Squares   F-statistic:                     213.1
Date:                Mon, 04 Dec 2017   Prob (F-statistic):           3.01e-73
Time:                        00:51:17   Log-Likelihood:                -524.01
No. Observations:                 714   AIC:                             1052.
Df Residuals:                     712   BIC:                             1061.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const               0          0        nan        nan           0           0
Age            0.0072      0.001     10.749      0.000       0.006       0.009
Fare           0.0033      0.000      9.289      0.000       0.003       0.004
==============================================================================
Omnibus:                      872.819   Durbin-Watson:                   1.876
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               63.317
Skew:                           0.365   Prob(JB):                     1.78e-14
Kurtosis:                       1.737   Cond. No.                          inf
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is      0. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

Age and Fare are not good predictors for surviving in the Titanic.

We have obtained bad models with linear regression. The R-Squared is 0.083 with a constant and 0.374 without a constant. Without a constant the model is so much better, but still really bad despite we have low p-values.